### DISCUSSION TITLE
Migrate files from nhost v1 to v2

### DISCUSSION THREAD

**Discussion by Svarto** [nhost/nhost#279] [Tips and Tricks]
Hi all,

This is for migrating user uploaded images (or other files) from nhost v1 to v2 keeping it all in memory and maintaining user ownership, it does the following:

1. Check if the file you want to migrate already exist in the destination (i.e. nhost v2)
2. If yes, skips to step 4, if not
3. Downloads the file from nhost v1 into memory
4. Still in memory, uploads the same file to nhost v2
5. Returns the file info in nhost v2 so that the reference can be inserted into your graphql database (e.g. images table or so).

Here is the code:
```
import axios from "axios";
import FormData from "form-data";
import { createClient, defaultExchanges } from "@urql/core";
import "isomorphic-unfetch";

// Migrate an image from nhost v1 to v2, maintaining
// user_id ownership and filename
export async function migrateImage({
  originImageURL,
  destinationBaseURL,
  destinationSecret,
  user_id,
  filename,
}) {
  const client_destination = createClient({
    url: `${destinationBaseURL}/v1/graphql`,
    exchanges: defaultExchanges,

    fetchOptions: () => ({
      headers: {
        ["x-hasura-admin-secret"]: destinationSecret,
      },
    }),
  });

  // Check if the filename already exists
  const { data: file_check, error } = await client_destination
    .query(FILE_CHECK, { _eq: filename })
    .toPromise();

  // If it does not exist, migrate the image
  if (file_check.files.length == 0) {
    const httpClient = axios.create({
      baseURL: `${destinationBaseURL}/v1/storage`,
    });

    // Download image from nhost v1
    const { data, headers } = await axios.get(originImageURL, {
      responseType: "stream",
    });

    const file = new FormData();
    file.append("file", data);

    // Upload image to destination nhost v2
    try {
      const { data: file_data } = await httpClient.post("/files", file, {
        headers: {
          ...file.getHeaders(),
          ["x-hasura-admin-secret"]: destinationSecret,
          ["x-hasura-role"]: "user",
          ["x-hasura-user-id"]: user_id,
          ["x-nhost-file-name"]: filename,
        },
      });

      // Return the uploaded file info to be further stored (e.g. in "user images" table in Hasura)
      return file_data;
    } catch (error) {
      console.error(error.message);
      return error;
    }
  } else {
    // If the filename already exist in storage, then return this files id and info
    return file_check.files[0];
  }
}

const FILE_CHECK = `
  query MyQuery($_eq: String) {
    files(where: {name: {_eq: $_eq}}) {
      id
      name
      createdAt
      updatedAt
      mimeType
      uploadedByUserId
    }
  }
  `;
```

Let me know if you have any questions.

**Comment by oristian** [comment]
This is helpful.  Thanks.   However, the challenge all of us will still have is that we need to manually find that URL of the image/file we'd like to migrate.  One suggestion could be to use  the old nhost_js_sdk + the admin secret to be able to call storage.getMetadata() to traverse the file structure we had set up over at nhost/minio and perform migration tasks in that way (for instance I have tied many files to a database row by using the primary key of the object and having a storage folder keyed by that uuid.  If a "hasura seeds create/apply" on that table will allow for the same UUID in the V2 instance, we can use that to associate those files to the same object in the V2 instance.  I've been trying to hack something together along these lines, but it's unclear how to use the V1 admin-secret in the Storage calls without the typical auth flow.  If you can help me figure that part out, I'd like to help to make this better.  Once the 1MB file storage limit is raised, I'm hoping to quickly get my photos moved over.  I can't even do it manually at this point without getting the CORS error.

I know nobody wants to build migration/conversion software that's only going to be used once

**Reply by Svarto** [reply]
Interesting challenge, you haven't stored the reference to your image in your Hasura DB? How is your frontend getting the image URL to display?

I simply pull my image DB rows from v1, compile the url for each image, and feed it into the above utility function and then I write it to V2 with another graphql mutation. 

The query and compile into URL is very similar to how my frontend displays images, the second part is just a graphql mutation.

I couldn't use the SDK as I wanted a script / node.js app I can just run from CLI without logging in. So i hit up the API directly.

Regarding the hasura seed / apply, i couldn't use that as my databas structure changed with the auth.user table breaking change in V2. I have built a couple of utility functions to help with DB migration I'll share later today or tomorrow.

A clarification, I'm not working at nhost, I'm just a customer like you and share your frustration in the migration. I am glad we can connect to leverage each other's code & approaches to lessen the migration pain.

**Reply by oristian** [reply]
Understood - I wasn't sure your role, but apologies if I was telegraphing any culpability your way.  Indeed I'm very frustrated with the current sub-V1 state, but eager to get on with it - this would only be more painful with more users and stakeholders wondering what happened.  I used a single column in my V1 table for sorting images and the folder they were in matched the UUID PK of the database row.  I used the getMetadata method to list the resources, but then reordered them in state when they were retrieved from the app.  It presented its own challenges but provided snappy reordering.

Luckily, this particular project of mine is super early days with just a handful of users ... I have currently assigned all my customer data to my V2 account, so that I can manage their images before assigning them back to new users I'll create soon.  In the meantime, I'm ironically using the V1 image transformation to get the images down below 1MB.  In post mortem feedback, I'll suggest that solid auth and storage migration tools should have been pre-requisites for a truly ready V2 beta.  But I prefer this scenario to one in which the promise of scalability might never have been realized in V1, with nowhere to go to be able to grow.

If I figure out a programmatic way to fetch the storageMetadata and and post the files to V2 with customizable mappings, I'll be sure to share my node script.  Thanks again for sharing your approach.
